src/pages/Account/Tabs/ModulesTab.tsx

```diff
@@ -179,6 +198,230 @@ function ViewCode({address}: {address: string}): JSX.Element {
   );
 }
 
+function WriteContract({address}: {address: string}) {
+  const {data, isLoading, error} = useGetAccountModules(address);
+  const [selectedModuleName, setSelectedModuleName] = useState<string>();
+  const [selectedFnName, setSelectedFnName] = useState<string>();
+
+  if (isLoading) {
+    return null;
+  }
+
+  if (error) {
+    return <Error address={address} error={error} />;
+  }
+
+  const modules = data ?? [];
+  if (modules.length === 0) {
+    return <EmptyTabContent />;
+  }
+
+  const moduleAndFnsGroup = modules.reduce((acc, module) => {
+    if (module.abi === undefined) {
+      return acc;
+    }
+
+    const fns = module.abi.exposed_functions.filter((fn) => fn.is_entry);
+    if (fns.length === 0) {
+      return acc;
+    }
+
+    const moduleName = module.abi.name;
+    return {
+      ...acc,
+      [moduleName]: fns,
+    } as Record<string, Types.MoveFunction[]>;
+  }, {} as Record<string, Types.MoveFunction[]>);
+
+  const module = modules.find((m) => m.abi?.name === selectedModuleName)?.abi;
+  const fn = selectedModuleName
+    ? moduleAndFnsGroup[selectedModuleName]?.find(
+        (fn) => fn.name === selectedFnName,
+      )
+    : undefined;
+
+  return (
+    <Grid container spacing={2}>
+      <Grid item md={3}>
+        <WriteContractSidebar
+          selectedModuleName={selectedModuleName}
+          selectedFnName={selectedFnName}
+          moduleAndFnsGroup={moduleAndFnsGroup}
+          handleClick={(moduleName: string, fnName: string) => {
+            setSelectedModuleName(moduleName);
+            setSelectedFnName(fnName);
+          }}
+        />
+      </Grid>
+      <Grid item md={9}>
+        {!module || !fn ? (
```

Comment:
```suggestion
        {!module ? (
```
no need to check for `fn` anymore? as long as `module` doesn't exist, `fn`'s def not selected; vice versa, as long as `module` is selected, we can display the form

Reply:
logically yes, however if we only check `!module ?` then `fn` has type `MoveFunction | undefined` which makes the code a bit difficult to read

the root cause is, we have not encoded the knowledge that "when fn is not null, then module must be not null as well" into TypeScript types.

We can do that though, by locating module and fn into just 1 type such as

```
type ModuleFn = {
  module: MoveModule;
  fn: MoveFunction;
}
```

and we do

```
const [moduleFn, setModuleFn] = useState<ModuleFn>();
```

wdyt? @jillxuu

Reply:
hmm seems like a overkill. i'm ok with the current implementation